### PR TITLE
feat(jenkins-agent) Add a docker build next to AWS and Azure (linux only)

### PR DESCRIPTION
**Pull request by smerle33**
add the docker (ubuntu-20) option to amazon and azure ones.

### CODE REVIEW

File: scripts/ubuntu-20-provision.sh [hunk 1]
```diff
@@ -37,163 +38,206 @@ function install_package_version() {
...
+  install_python
+  install_git_gitlfs
+  install_jdk
+  install_docker_compose
+  install_maven
+  setuser
+  cleanup
+}
+
+main
```
Comment (dduportal) [inline scripts/ubuntu-20-provision.sh:243]:
```suggestion
main

```

**Review by dduportal — APPROVED** [review]

File: Jenkinsfile_k8s [hunk 2]
```diff
@@ -94,11 +94,21 @@ pipeline {
               values 'amazon-ebs'
             }
           }
+          // No build on windows / docker not yet implemented
```
Comment (lemeurherve) [inline Jenkinsfile_k8s:97]:
```suggestion
          // No build on Windows, Docker not yet implemented
```

File: jenkins-agent.pkr.hcl [hunk 3]
```diff
@@ -203,6 +225,18 @@ build {
     vm_size         = local.azure_vm_size
   }
 
+  provisioner "shell" {
+    # steps for docker images
+    only = ["docker.ubuntu-20"]
+    environment_vars = [
+      "DEBIAN_FRONTEND=noninteractive",
+    ]
+    #adding sudo and bash for docker software-properties-common
```
Comment (lemeurherve) [inline jenkins-agent.pkr.hcl:234]:
```suggestion
    # adding sudo and bash for docker software-properties-common 
```
I restreined myself asking for an uppercase at the beginning of all comments

File: run-packer.sh [hunk 4]
```diff
@@ -3,17 +3,51 @@
...
 
+packer_cmd="packer"
 packer_template_dir="./"
+packer_install_dir="/$HOME"
+packer_version="1.7.8"
 
 export PKR_VAR_scm_ref PKR_VAR_image_type PKR_VAR_agent
 
 
+## check if packer exist or install it
```
Comment (lemeurherve) [inline run-packer.sh:18]:
```suggestion
## check if packer exists or install it
```

File: run-packer.sh [hunk 5]
```diff
@@ -3,17 +3,51 @@
...
+if ! command -v ${packer_cmd} >/dev/null 2>&1
+then
+  if test -x "${packer_install_dir}/${packer_cmd}"
+  then
+    packer_cmd="${packer_install_dir}/packer"
+  else
+    echo "Packer not installed, I install it"
```
Comment (lemeurherve) [inline run-packer.sh:25]:
```suggestion
    echo "Packer not installed, installing it"
```

File: scripts/ubuntu-20-provision.sh [hunk 6]
```diff
@@ -7,19 +7,20 @@ set -eux -o pipefail
...
-## Check for presence of requirements or fail fast
-for cli in add-apt-repository apt-get apt-cache awk curl grep groupadd head tar uname useradd
-do
-  if ! command -v $cli >/dev/null 2>&1
-  then
-    echo "ERROR: command line ${cli} required but not found. Exiting."
-    exit 1
-  fi
-done
+## This function check a list of commands are working. exit with code 1 if not
```
Comment (lemeurherve) [inline scripts/ubuntu-20-provision.sh:12]:
```suggestion
## This function checks a list of commands are working, and exits with code 1 if not
```

File: scripts/ubuntu-20-provision.sh [hunk 7]
```diff
@@ -37,163 +38,206 @@ function install_package_version() {
...
+  cp /tmp/add_auth_key_to_user.sh /usr/local/bin/add_auth_key_to_user.sh
+  chmod a+x /usr/local/bin/add_auth_key_to_user.sh
+}
 
-## Disable and Remove Unattended APT Upgrades
-echo 'APT::Periodic::Enable "0";' > /etc/apt/apt.conf.d/10cloudinit-disable
-apt-get purge -y unattended-upgrades || true # Do not fail if the package does not exist
+## All the clean for apt
+function clean_apt() {
+  ## Disable and Remove Unattended APT Upgrades
```
Comment (lemeurherve) [inline scripts/ubuntu-20-provision.sh:48]:
```suggestion
  ## Disable and remove Unattended APT upgrades
```

File: scripts/ubuntu-20-provision.sh [hunk 8]
```diff
@@ -37,163 +38,206 @@ function install_package_version() {
...
+    qemu-user-static
+
+  if [ "$(uname -m)" = "x86_64" ]; then
+    docker run --rm --privileged multiarch/qemu-user-static --reset -p yes || true # Never fail
+  fi
+}
+
+## Install Python 3
+function install_python() {
+  # Install Python 3
```
Comment (lemeurherve) [inline scripts/ubuntu-20-provision.sh:103]:
```suggestion
```

File: scripts/ubuntu-20-provision.sh [hunk 9]
```diff
@@ -37,163 +38,206 @@ function install_package_version() {
...
+function install_JA_requirements(){
+  apt-get install -y --no-install-recommends \
+    make \
+    unzip \
+    zip \
+    jq \
+    parallel
+}
 
 # setup qemu
```
Comment (lemeurherve) [inline scripts/ubuntu-20-provision.sh:89]:
```suggestion
## setup qemu
```
Not sure why the use of `#` or `##`

File: scripts/ubuntu-20-provision.sh [hunk 10]
```diff
@@ -37,163 +38,206 @@ function install_package_version() {
...
+  # Install Python 3
+  apt-get install -y --no-install-recommends \
+    python3 \
+    python3-docker \
+    python3-pip \
+    python3-venv \
+    python3-wheel
+}
+
+## Install git and gitLFS
```
Comment (lemeurherve) [inline scripts/ubuntu-20-provision.sh:112]:
```suggestion
## Install git and git-lfs
```
For consistency and ease of search

**Comment by lemeurherve** [comment]
@dduportal @smerle33 minor comments on comments, you might want to disable the auto-merge if you want to take them in account.

File: scripts/ubuntu-20-provision.sh (hunk 9 above)
Comment (dduportal) [reply scripts/ubuntu-20-provision.sh:89]: I tend to use double by default, to allow "simple" for sub sections.

**Comment by dduportal** [comment]
Last build was all green: let's merge! (the gihub check are marked as failing because of build #30 while #31 worked very well)

**Pull request merged**